File: backend/app/schemas/concept_graph.py
"""Pydantic schemas for concept graph API."""
from typing import Literal

from pydantic import BaseModel


class GraphNode(BaseModel):
    id: int
    user_id: int
    course_id: int
    knowledge_point_id: int | None = None
    title: str
    normalized_title: str
    summary: str = ""
    aliases: list[str] = []
    importance: int = 3
    source_chunk_ids: list[int] = []
    weak_point_score: float = 0.0


class GraphEdge(BaseModel):
    id: int
    user_id: int
    source_node_id: int
    target_node_id: int
    relation_type: str
    confidence: float
    reason: str = ""
    evidence_chunk_ids: list[int] = []
    status: str = "candidate"
    audit_run_id: int | None = None


class GraphResponse(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]


class RebuildResponse(BaseModel):
    nodes_count: int
    edges_count: int


class NodeDetailResponse(GraphNode):
    related_edges: list[GraphEdge] = []


class EdgeActionResponse(GraphEdge):
    pass


class CompareRequest(BaseModel):
    source_node_id: int
    target_node_id: int
    edge_id: int | None = None
    user_focus: Literal["concept", "exam", "transfer"] = "concept"


class CompareReportResponse(BaseModel):
    id: int
    source_node_id: int
    target_node_id: int
    edge_id: int | None = None
    report_json: dict
    citation_chunk_ids: list[int] = []
    prompt_version: str = "v1"
    provider: str = "mock"
    model_name: str = "mock"
    fallback_used: bool = False
    fallback_reason: str = ""
    audit_run_id: int | None = None

File: frontend/src/api/conceptGraph.ts
import request from './index'
import type { AxiosPromise } from 'axios'

export interface GraphNode {
  id: number
  user_id: number
  course_id: number
  knowledge_point_id: number | null
  title: string
  normalized_title: string
  summary: string
  aliases: string[]
  importance: number
  source_chunk_ids: number[]
  weak_point_score: number
}

export interface GraphEdge {
  id: number
  user_id: number
  source_node_id: number
  target_node_id: number
  relation_type: string
  confidence: number
  reason: string
  evidence_chunk_ids: number[]
  status: string
  audit_run_id: number | null
}

export interface GraphResponse {
  nodes: GraphNode[]
  edges: GraphEdge[]
}

export interface RebuildResponse {
  nodes_count: number
  edges_count: number
}

export interface NodeDetail extends GraphNode {
  related_edges: GraphEdge[]
}

export interface CompareReport {
  id: number
  source_node_id: number
  target_node_id: number
  edge_id: number | null
  report_json: Record<string, any>
  citation_chunk_ids: number[]
  prompt_version: string
  provider: string
  model_name: string
  fallback_used: boolean
  fallback_reason: string
  audit_run_id: number | null
}

export function rebuildGraph(): AxiosPromise<RebuildResponse> {
  return request.post('/concept-graph/rebuild')
}

export function getGraph(params?: {
  course_ids?: string
  relation_type?: string
  status?: string
}): AxiosPromise<GraphResponse> {
  return request.get('/concept-graph', { params })
}

export function getNodeDetail(nodeId: number): AxiosPromise<NodeDetail> {
  return request.get(`/concept-graph/nodes/${nodeId}`)
}

export function confirmEdge(edgeId: number): AxiosPromise<GraphEdge> {
  return request.post(`/concept-graph/edges/${edgeId}/confirm`)
}

export function rejectEdge(edgeId: number): AxiosPromise<GraphEdge> {
  return request.post(`/concept-graph/edges/${edgeId}/reject`)
}

export function compareNodes(
  sourceNodeId: number,
  targetNodeId: number,
  edgeId?: number,
  userFocus: string = 'concept',
  forceRefresh = false,
): AxiosPromise<CompareReport> {
  // The compare endpoint calls the user's LLM which can take 50-90s.
  // Override the default 30s axios timeout so the request isn't
  // aborted before the backend finishes generating and caching the
  // report (which would leave the UI showing "暂无对比报告").
  return request.post(
    '/concept-graph/compare',
    {
      source_node_id: sourceNodeId,
      target_node_id: targetNodeId,
      edge_id: edgeId ?? null,
      user_focus: userFocus,
      force_refresh: forceRefresh,
    },
    { timeout: 120000 },
  )
}
